### PR TITLE
Defunctionalization with base AST

### PR DESCRIPTION
**Pull request by HarrisL2**

### CODE REVIEW

File: compiler/shared/main/scala/mlscript/compiler/mono/specializer/BoundedTerm.scala [hunk 1]
```diff
@@ -0,0 +1,180 @@
+package mlscript.compiler.mono
+import mlscript.{Var, Lit}
+import scala.collection.mutable.{Map => MutMap}
+import scala.collection.immutable.ListMap
+
+
+abstract class MonoVal extends MonoValImpl
```
Comment (LPTK) [inline compiler/shared/main/scala/mlscript/compiler/mono/specializer/BoundedTerm.scala]: Should be sealed?!

File: compiler/shared/main/scala/mlscript/compiler/mono/specializer/BoundedTerm.scala [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+  override def toString: String = this match {
+    case FuncVal(name, params, ctx) => s"FuncVal(${name}, ${params.map(_.mkString("(",",",")")).getOrElse("None")}, ${ctx})"
+    case LiteralVal(i) => s"LiteralVal(${i})"
+    case ObjVal(name, params, fields) => s"ObjVal(${name}, ${fields})"
+    case TypeVal(name) => s"TypeVal(${name})"
+    case TupVal(fields) => s"TupVal(${fields})"
+    case UnknownVal() =>  s"UnknownVal"
+    case PrimVal() => s"PrimVal()"
+    case VarVal(vx, version, _) => s"VarVal(${vx})"
+      }
```
Comment (LPTK) [inline compiler/shared/main/scala/mlscript/compiler/mono/specializer/BoundedTerm.scala]:
```suggestion
  }
```

File: compiler/shared/test/diff/Defunctionalize/ObjFieldAccess.mls [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+//│ fun get1$A$1: Object -> "6"
+//│ fun get2$A$1: Object -> "6"
+//│ fun main$$3: () -> "6"
+//│ fun main$$2: () -> "6"
+//│ let a$1: A$1
+//│ "6"
+//│ Code generation encountered an error:
+//│   unguarded recursive use of by-value binding a$1
+
+:ge //TODO: top-level defs
```
Comment (LPTK) [inline compiler/shared/test/diff/Defunctionalize/ObjFieldAccess.mls]: Please be a bit more descriptive.

File: compiler/shared/test/diff/Defunctionalize/OldMonoList.mls [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+//│ fun main$$5: () -> List$1
+//│ fun apply$Lambda1$3$4: (anything, Int) -> Int
+//│ fun map$Nil$2: forall 'a. ('a & (List$1 | Nil$2), anything) -> (Nil$2 | 'a)
+//│ fun apply$Lambda1$2$3: (anything, Int) -> Int
+//│ List$1
+//│ res
+//│     = List$1 {}
+
+:mono
+:AllowRuntimeErrors
```
Comment (LPTK) [inline compiler/shared/test/diff/Defunctionalize/OldMonoList.mls]: This is file-global.

File: compiler/shared/test/diff/Defunctionalize/RecursiveFunc.mls [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+//│ TypingUnit {
+//│   class List$1([val l: |(|(List$1, Nil$2,), undefined,), val hasTail: Bool,]) {}
+//│   class Nil$2([val l: |(|(List$1, Nil$2,), undefined,), val hasTail: Bool,]) {}
+//│   let l$2 = (lst).l
+//│   fun count$1 = (lst,) => {if ((lst).hasTail) then {if (is(l$2, undefined,)) then 1 else +(count$1(l$2,), 1,)} else 0}
+//│   Code(List(count$1((new List$1)((new List$1)((new Nil$2)(undefined, false,), true,), true,),)))
+//│ }
+//│ Mono:
+//│ Variable lst not found during evaluation
+//│ ╔══[ERROR] Post-process failed to produce AST, defaulting to standard AST
```
Comment (LPTK) [inline compiler/shared/test/diff/Defunctionalize/RecursiveFunc.mls]: This should be a proper "unimplemented" crash with a stack trace.

File: compiler/shared/main/scala/mlscript/compiler/Helpers.scala [hunk 6]
```diff
@@ -3,176 +3,84 @@ package compiler
...
+import mlscript.IntLit
+import mlscript.StrLit
+import mlscript.AppliedType
+import mlscript.TypeName
+import mlscript.TypeDefKind
+import mlscript.compiler.mono.Monomorph
+import mlscript.compiler.NuParameter
+import mlscript.NuDecl
+import mlscript.TypingUnit
+import mlscript.Located
```
Comment (LPTK) [inline compiler/shared/main/scala/mlscript/compiler/Helpers.scala]: Fix import slab.

File: compiler/shared/test/diff/Defunctionalize/SimpleFunc.mls [hunk 7]
```diff
@@ -54,12 +54,16 @@ let b = foo(23, false)
 //│ }
 //│ Mono:
 //│ Bounds for Var(foo$3) not found during defunc.
-//│ ╔══[ERROR] Post-process failed to produce AST, defaulting to standard AST
```
Comment (LPTK) [inline compiler/shared/test/diff/Defunctionalize/SimpleFunc.mls]: You should change:
```diff
-:e // TODO: Evaluate unused terms
+// TODO: Evaluate unused terms
```
to remove the stack trace from the test file.

Comment (LPTK) [reply compiler/shared/test/diff/Defunctionalize/SimpleFunc.mls]: Oh, is this a `MonomorphError`, whose handling is special-cased below? Why the special case code?

Comment (HarrisL2) [reply compiler/shared/test/diff/Defunctionalize/SimpleFunc.mls]: I don't think there's any particular reason other than to distinguish it from the other sections of the codebase. Would it be appropriate to change this to a `CodeGenError` and use the same error handling system?

Comment (HarrisL2) [reply compiler/shared/test/diff/Defunctionalize/SimpleFunc.mls]: Added conditional that only outputs stack trace if the `:ge` flag is given

**Pull request merged**